packages/agent/src/oidc.ts

```diff
@@ -0,0 +1,703 @@
+      name : 'HKDF',
+      hash : 'SHA-256',
+      info : new Uint8Array(),
+      salt : new Uint8Array(),
+    },
+    derivedKey,
+    256
+  );
+  const sharedEncryptionKey = new Uint8Array(derivedKeyBits);
+  return sharedEncryptionKey;
+}
+
+function encryptAuthResponse({
+  jwt,
+  encryptionKey,
+  providerDidKid,
+  randomPin,
+}: {
+  jwt: string;
+  encryptionKey: Uint8Array;
+  providerDidKid: string;
+  randomPin: string;
+}) {
+  const protectedHeader = {
+    alg : 'dir',
+    cty : 'JWT',
+    enc : 'XC20P',
+    typ : 'JWT',
+    kid : providerDidKid,
+  };
+  const nonce = CryptoUtils.randomBytes(24);
+  const additionalData = Convert.object({
+    ...protectedHeader,
+    pin: randomPin,
+  }).toUint8Array();
+
+  const jwtBytes = Convert.string(jwt).toUint8Array();
+  const chacha = xchacha20poly1305(encryptionKey, nonce, additionalData);
+  const ciphertextAndTag = chacha.encrypt(jwtBytes);
+
+  /** The cipher output concatenates the encrypted data and tag
+   * so we need to extract the values for use in the JWE. */
+  const ciphertext = ciphertextAndTag.subarray(0, -16);
+  const authenticationTag = ciphertextAndTag.subarray(-16);
+
+  const compactJwe = [
+    Convert.object(protectedHeader).toBase64Url(),
+    '', // Empty string since there is no wrapped key.
+    Convert.uint8Array(nonce).toBase64Url(),
+    Convert.uint8Array(ciphertext).toBase64Url(),
+    Convert.uint8Array(authenticationTag).toBase64Url(),
+  ].join('.');
+
+  return compactJwe;
+}
+
+export async function createPermissionGrants(
+  selectedDids: string[],
+  delegateDid: BearerDid,
+  dwn: AgentDwnApi
+) {
+  const grantPromises = selectedDids.map(async (did) => {
+    // TODO: Replace with real permission request
+    const permissionRequestData = {
+      description : 'The app is asking to Records Write to http://profile-protocol.xyz',
+      scope       : {
+        interface : DwnInterfaceName.Records,
+        method    : DwnMethodName.Write,
+        protocol  : 'http://profile-protocol.xyz',
+      },
+    };
+
+    // TODO: Confirm this
+    const message = await dwn.processRequest({
+      author        : did,
+      target        : did,
+      messageType   : DwnInterface.RecordsWrite,
+      messageParams : {
+        recipient    : delegateDid.uri,
+        protocolPath : 'grant',
+        protocol     : ' https://tbd.website/dwn/permissions',
+        dataFormat   : 'application/json',
+        data         : Convert.object(permissionRequestData).toUint8Array(),
+      },
+      // todo: is it data or datastream?
+      // dataStream: await Convert.object(permissionRequestData).toBlobAsync(),
+    });
+
+    return message;
+  });
+
+  const permissionGrants = (await Promise.all(grantPromises)).flat();
+
+  return permissionGrants;
+}
+
+async function submitAuthResponse(
+  selectedDids: string[],
+  authRequest: Web5ConnectAuthRequest,
+  randomPin: string,
+  dwn: AgentDwnApi
+) {
+  const delegateDid = await DidDht.create();
```

Comment:
The delegateDID should be a `did:jwk`, there's no need for it to have service endpoints or key-rolling.

Reply:
I agree and thats what I was doing at first (did:key or did:jwk) but frank recommended I use DHT for everything so now would be a rewrite to go back again. @thehenrytsai what do you think?

Reply:
I think Frank may have meant the Agent DIDs, because those actually do need to sync. We should transition the permissioned DIDs given to apps to be DID JWK.

Reply:
he really did mean all dids but i might have taken what he was saying too literally and this is an exception

Reply:
Agree with @csuwildcat and @LiranCohen that the app only needs a DID that essentially holds a private key. This is the reason I questioned why this was originally an array of private keys instead of one in my original comment.

I am personally fine with any DID method as long as it does not require resolution in this use case. I care more about agreeing with the actual structure we are stamping as the standard for shipping the DID (with private key), and looks like `PortableDid` is the candidate of choice.

Reply:
ok so `did:jwk` with PortableDid?

have we abandoned `did:key` or whats the usecase for it anymore?

Reply:
fixed